extensions-utils/src/dfx.rs

```diff
@@ -50,6 +54,57 @@ where
         })
 }
 
+/// Calls a binary that was delivered with an extension tarball.
+///
+/// # Returns
+/// - On success, returns stdout as a string.
+/// - On error, returns an error message including stdout and stderr.
+#[context("Calling {} CLI failed, or, it returned an error.", binary_name)]
+pub fn call_extension_bundled_binary<S, I>(binary_name: &str, args: I) -> anyhow::Result<String>
+where
+    I: IntoIterator<Item = S>,
+    S: AsRef<OsStr>,
+{
+    let extension_binary_path =
+        std::env::current_exe().map_err(|e| anyhow::anyhow!("Failed to get current exe: {}", e))?;
+    let extension_dir_path = extension_binary_path
+        .parent()
+        .unwrap_or_else(|| Path::new("."));
+    let binary_to_call = extension_dir_path.join(binary_name);
+
+    let mut command = Command::new(&binary_to_call);
+    command.args(args);
+    // The sns command line tool itself calls dfx; it should call this dfx.
+    // The sns command line tool should not rely on commands not packaged with dfx.
+    // The same applies to other bundled binaries.
+    command.env(
+        "PATH",
+        binary_to_call.parent().unwrap_or_else(|| Path::new(".")),
+    );
+    command
+        .stdin(process::Stdio::null())
+        .output()
+        .map_err(anyhow::Error::from)
+        .and_then(|output| {
+            if output.status.success() {
+                Ok(String::from_utf8_lossy(&output.stdout).into_owned())
+            } else {
+                let args: Vec<_> = command
```

Comment:
Please take a look at the '"npm is not installed"'` test in https://github.com/dfinity/sdk/blob/master/e2e/tests-dfx/error_context.bash#L127

Change `assert_contains 'program: "npm"'` to check for something else, or otherwise take a look at the error output.

The rust 1.71.1 update brought along pretty detailed `Command` output.  It would be worthwhile to see if we can make use of that here, without rolling our own.

Reply:
this one is still TODO